### PR TITLE
Merge Develop

### PR DESCRIPTION
**Pull request by ercanozkaya**

### CODE REVIEW

**Comment by amazeika** [comment]
@ercanozkaya This is triggering a bit too late

https://github.com/joomlatools/joomlatools-framework/blob/674cc29fca2184edd0775018b15d4b151a9d47fa/code/libraries/joomlatools/library/dispatcher/http.php#L134

We now have ComKoowaDispatcherBehaviorDecoratable which is kicking before dispatch and gets to grab a controller instance here

https://github.com/joomlatools/joomlatools-framework/blob/674cc29fca2184edd0775018b15d4b151a9d47fa/code/libraries/joomlatools/component/koowa/dispatcher/behavior/decoratable.php#L128

At this point the controller name based on the current view isn't set.

I think that controller should be properly set before, probably at the time we instantiate the dispatcher. @johanjanssens thoughts?

**Comment by amazeika** [comment]
@ercanozkaya I see that we continue to support tmpl=koowa. If present decoratable assumes that koowa is the decorator. This is ok.

However there's a problem when loading jQuery. The decorator isn't set on the template which makes the template helper to assume that the decorator is joomla, so it loads kquery on the Joomla document:

https://github.com/joomlatools/joomlatools-framework/blob/develop/code/libraries/joomlatools/component/koowa/template/helper/behavior.php#L111-L123

since dispatcher thinks that decorator is koowa, then it terminates the request. kQuery is therefore never loaded. 

Decoratable should perhaps set the layout or the view decorator to koowa if tmpl=koowa is the query. thoughts?

**Comment by amazeika** [comment]
@ercanozkaya elaborating the idea above, I was thinking on something like

`    protected function _beforeDispatch(KDispatcherContextInterface $context)
    {
        $request  = $this->getRequest();

        if($request->getQuery()->tmpl === 'koowa') {
            $this->getController()->getView()->setDecorator('koowa');
        }
...`

on ComKoowaDispatcherBehaviorDecoratable. Basically moving it out from getDecorator so you get:

`public function getDecorator()
    {
        $request  = $this->getRequest();
        $response = $this->getResponse();

        if($request->getHeaders()->has('X-Flush-Response') || $response->isError()) {
            $result = 'koowa';
        } else {
            $result = $this->getController()->getView()->getDecorator();
        }

        return $result;
    }`

would that work or break something?

**Comment by amazeika** [comment]
@ercanozkaya As discussed on slack, I'll proceed to specifying the decorator on each view. Trying to set it on the dispatcher (see above) is unreliable.

Should we remove the tmp=koowa check on ComKoowaDispatcherBehaviorDecoratable::getDecorator? As discussed these queries with tmp=koowa will break if they have script assets added on their layouts (they get pushed to the Joomla document).

**Comment by johanjanssens** [comment]
@amazeika The idea is that the decorator is defined per view, not per dispatcher, doing it on the dispatcher will not work well, the dispatcher doesn't understand what it's dispatching too. 

@amazeika  @ercanozkaya  BC is left in place as much as possible to make code migration easier, the goal is that each extension is refactored to use the new approach and the BC code is then removed. 

Does that make sense?

File: code/libraries/joomlatools/library/dispatcher/http.php [hunk 1]
```diff
@@ -100,6 +100,21 @@ protected function _loadTranslations(KControllerContextInterface $context)
         $this->getObject('translator')->load($identifier);
     }
 
+    protected function _beforeDispatch(KDispatcherContextInterface $context)
```
Comment (johanjanssens) [inline code/libraries/joomlatools/library/dispatcher/http.php]: @amazeika @ercanozkaya This is good, same approach i did in Kodekit. See: https://github.com/timble/kodekit/blob/master/code/dispatcher/behavior/routable.php#L28 We can leave this here, don't need to move it to a behavior, change doesn't need syncing.

**Comment by amazeika** [comment]
> @amazeika The idea is that the decorator is defined per view, not per dispatcher, doing it on the dispatcher will not work well, the dispatcher doesn't understand what it's dispatching too.
> 
> @amazeika @ercanozkaya BC is left in place as much as possible to make code migration easier, the goal is that each extension is refactored to use the new approach and the BC code is then removed.
> 
> Does that make sense?

Agreed

**Pull request merged**